pkg/retriever/bitswaphelpers/multiblockstore.go

```diff
@@ -0,0 +1,157 @@
+	}
+}
+
+// AddLinkSystem registers a linksystem to use for a given retrieval id
+func (mbs *MultiBlockstore) AddLinkSystem(id types.RetrievalID, lsys *linking.LinkSystem) error {
+	mbs.linkSystemsLk.Lock()
+	defer mbs.linkSystemsLk.Unlock()
+	if _, ok := mbs.linkSystems[id]; ok {
+		return ErrAlreadyRegisterd
+	}
+	mbs.linkSystems[id] = lsys
+	return nil
+}
+
+// RemoveLinkSystem unregisters the link system for a given retrieval id
+func (mbs *MultiBlockstore) RemoveLinkSystem(id types.RetrievalID) {
+	mbs.linkSystemsLk.Lock()
+	defer mbs.linkSystemsLk.Unlock()
+	delete(mbs.linkSystems, id)
+}
+
+// DeleteBlock is not supported
+func (mbs *MultiBlockstore) DeleteBlock(ctx context.Context, c cid.Cid) error {
+	return ErrNotSupported
+}
+
+// Has is not supported
+func (mbs *MultiBlockstore) Has(ctx context.Context, c cid.Cid) (bool, error) {
+	return false, ErrNotSupported
+}
+
+type byteReader interface {
+	Bytes() []byte
+}
+
+// Get returns a block only if the given ctx contains a retrieval ID as a value that
+// references a known linksystem. If it does, it uses that linksystem to load the block
+func (mbs *MultiBlockstore) Get(ctx context.Context, c cid.Cid) (blocks.Block, error) {
+	id, err := types.RetrievalIDFromContext(ctx)
+	if err != nil {
+		if errors.Is(err, types.ErrMissingContextKey) {
+			return nil, format.ErrNotFound{Cid: c}
+		}
+		return nil, err
+	}
+	mbs.linkSystemsLk.RLock()
+	lsys, ok := mbs.linkSystems[id]
+	mbs.linkSystemsLk.RUnlock()
+	if !ok {
+		return nil, format.ErrNotFound{Cid: c}
+	}
+	r, err := lsys.StorageReadOpener(linking.LinkContext{Ctx: ctx}, cidlink.Link{Cid: c})
+	if err != nil {
+		return nil, err
+	}
+	if br, ok := r.(byteReader); ok {
+		return blocks.NewBlockWithCid(br.Bytes(), c)
+	}
+	data, err := io.ReadAll(r)
+	if err != nil {
+		return nil, err
+	}
+	return blocks.NewBlockWithCid(data, c)
+}
+
+// GetSize is unsupported
+func (mbs *MultiBlockstore) GetSize(ctx context.Context, c cid.Cid) (int, error) {
+	return 0, errors.New("not supported")
+}
+
+// Put writes a block only if the given ctx contains a retrieval ID as a value that
+// references a known linksystem. If it does, it uses that linksystem to save the block
+func (mbs *MultiBlockstore) Put(ctx context.Context, blk blocks.Block) error {
+	id, err := types.RetrievalIDFromContext(ctx)
+	if err != nil {
+		if errors.Is(err, types.ErrMissingContextKey) {
+			return ErrNotSupported
+		}
+		return err
+	}
+	mbs.linkSystemsLk.RLock()
+	lsys, ok := mbs.linkSystems[id]
+	mbs.linkSystemsLk.RUnlock()
+	if !ok {
+		return ErrNotSupported
+	}
+	w, commit, err := lsys.StorageWriteOpener(linking.LinkContext{Ctx: ctx})
+	if err != nil {
+		return err
+	}
+	_, err = w.Write(blk.RawData())
+	if err != nil {
+		return err
+	}
+	return commit(cidlink.Link{Cid: blk.Cid()})
+}
+
+// PutMany puts a slice of blocks at the same time, with the same rules as Put
+func (mbs *MultiBlockstore) PutMany(ctx context.Context, blks []blocks.Block) error {
```

Comment:
Any idea whether this is used in Bitswap? If so, maybe we should swap them around and do all the storage logic in here and call it from `Put()` so we don't end up doing all the same work for each block. In fact, maybe we should anyway as long as we're supporting this.

Reply:
right.